### PR TITLE
Finalize super_text package for release (Resolves #520)

### LINKED ISSUE #520: Change SuperSelectableText to use layers for multi-user selections
Change SuperSelectableText to use layers for multi-user selections.

Superlist needs to display multiple user selections at the same time. To prepare to support this ability, define a `SuperText` widget that displays layers above and below the text, and then use `SuperText` to implement `SuperSelectableText`.

### PR DESCRIPTION
**Pull request by matthew-carroll**
Finalizes the `super_text` package for release.

We knew we wanted to make some changes to `SuperSelectableText`, and those changes needed to be made before we released `super_text`, otherwise we'd immediately break all the APIs.

This PR eliminates the `SuperSelectableText` widget. In its place is a new widget called `SuperTextWithSelection`.

The new `SuperTextWithSelection` has a `.single()` constructor for traditional selection painting. It also has `.multi()` for multi-user selections.

https://user-images.githubusercontent.com/7259036/162927869-551df577-7bc9-4b7a-b996-894fbc1b2970.mov

`SuperTextWithSelection` is built on top of `SuperText`, which is a widget that displays anything you want above its text, and beneath its text. With the introduction of `SuperText`, anyone can build their own custom text selections and decorations.

https://user-images.githubusercontent.com/7259036/162927789-394ffa79-9bcc-4996-9dc6-009d069d1236.mov

At the time of putting up this PR, there are at least a couple issues that I still need to resolve. However, 99% of what's here is intended to merge, so I thought it would be good to get started with a review.

### CODE REVIEW

**Review by roughike — COMMENTED** [review]
The `SuperTextWithSelection` & `UserSelection` API seems exactly what we need.

I didn't find how this works in the context of `SuperEditor` & `DocumentSelection` though - is there something I missed?

**Comment by matthew-carroll** [comment]
@roughike the multi-user capabilities haven't been extended to documents yet. The only changes in super_editor that this PR includes, are the changes that were compile-time necessary to keep it working with super_text. Once super_text is released, we'll tackle the document level. 

When we get to documents, we'll need to chat about what makes sense from an API standpoint. For example, do we really want a single editor API that covers single-selection and multi-selection? Or would it be more appropriate to branch the editor implementation in two different directions at that point? We can have those conversations when we get there.

File: super_text/lib/src/super_text_with_selection.dart [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+  Widget _buildLayerAbove(BuildContext context, TextLayout textLayout) {
+    buildsLog.info("Building SuperTextWithSelection ($hashCode) caret layer");
+    return ValueListenableBuilder<List<UserSelection>>(
+      valueListenable: widget.userSelections,
+      builder: (context, value, child) {
+        buildsLog.info(
+            "SuperTextWithSelection ($hashCode) user selection changed, building carets: ${widget.userSelections.value.isNotEmpty ? widget.userSelections.value.first.selection : "null"}");
+        return Stack(
+          children: [
+            for (final userSelection in widget.userSelections.value)
```
Comment (miguelcmedeiros) [inline super_text/lib/src/super_text_with_selection.dart:216]: Same as above

**Review by miguelcmedeiros — COMMENTED** [review]
Overall looks good/understandable 👍 

Will probably do another pass once you wrap the WIP. Left a few nits.

File: super_text/lib/src/infrastructure/fill_width_if_constrained.dart [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import 'package:flutter/rendering.dart';
+import 'package:flutter/widgets.dart';
+
+/// Forces [child] to take up all available width when the
+/// incoming width constraint is bounded, otherwise the [child]
+/// is sized by its intrinsic width.
+///
+/// If there is an existing widget that does this, get rid of this
+/// widget and use the standard widget.
+class FillWidthIfConstrained extends SingleChildRenderObjectWidget {
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_text/lib/src/infrastructure/fill_width_if_constrained.dart:10]: I don't think so

File: super_text/lib/src/super_text_with_selection.dart [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+  Widget _buildLayerBeneath(BuildContext context, TextLayout textLayout) {
+    buildsLog.info("Building SuperTextWithSelection ($hashCode) selection highlight layer");
+    return ValueListenableBuilder<List<UserSelection>>(
+      valueListenable: widget.userSelections,
+      builder: (context, value, child) {
+        buildsLog.info(
+            "SuperTextWithSelection ($hashCode) user selection changed, building selection highlights: ${widget.userSelections.value.isNotEmpty ? widget.userSelections.value.first.selection : "null"}");
+        return Stack(
+          children: [
+            for (final userSelection in widget.userSelections.value)
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply super_text/lib/src/super_text_with_selection.dart:183]: I don't think it really matters, but I updated the two references.

**Comment by matthew-carroll** [comment]
@miguelcmedeiros @roughike - I think we're ready for another review on this. @slightfoot and I were able to get `SuperText` to layout text and decorations in a single frame. This means that we shouldn't have to worry about the underlying `RenderParagraph` every silently complaining that the layout is invalid, and we shouldn't ever see a caret or text selection fall a frame behind the text layout.

**Review by miguelcmedeiros — APPROVED** [review]
LGTM

**Comment by matthew-carroll** [comment]
There are 2 failing tests. Both of them are mobile auto-scroll tests, one for Android and one for iOS:

```
SuperEditor auto-scroll on Android, keeps caret visible when keyboard appears
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: a value greater than or equal to <490.0>
  Actual: <464.79999923706055>
   Which: is not a value greater than or equal to <490.0>
```

I'm going to merge this PR with the failing tests so that we can move forward.

@ThangVuNguyenViet can you look into this again?

**Comment by ThangVuNguyenViet** [comment]
@matthew-carroll Sure I will

**Pull request merged**